src/core/ucc_team.c

```diff
@@ -229,3 +237,110 @@ ucc_status_t ucc_team_destroy(ucc_team_h team)
     }
     return status;
 }
+
+ucc_status_t ucc_ee_create(ucc_team_h team, const ucc_ee_params_t *params,
+                           ucc_ee_h *ee_p)
+{
+    ucc_ee_t *ee;
+
+    ee = ucc_malloc(sizeof(ucc_ee_t), "ucc execution engine");
+    if (!ee) {
+        ucc_error("failed to allocate %zd bytes for ucc execution engine",
+                  sizeof(ucc_ee_t));
+        return UCC_ERR_NO_MEMORY;
+    }
+
+    ee->team = team;
+    ee->ee_type = params->ee_type;
+    ee->ee_context_size = params->ee_context_size;
+    ee->ee_context = params->ee_context;
+    ucc_spinlock_init(&ee->lock, 0);
+    ucc_queue_head_init(&ee->event_in_queue);
+    ucc_queue_head_init(&ee->event_out_queue);
+    *ee_p = ee;
+
+    ucc_info("ucc_ee_create: ee is created ");
+
+    return UCC_OK;
+}
+
+ucc_status_t ucc_ee_destroy(ucc_ee_h ee)
+{
+    ucc_spinlock_destroy(&ee->lock);
+    ucc_free(ee);
+
+    ucc_info("ucc_ee_create: ee is destroyed ");
+
+    return UCC_OK;
+}
+
+ucc_status_t ucc_ee_get_event_internal(ucc_ee_h ee, ucc_ev_t **ev, ucc_queue_head_t *queue)
+{
+    ucc_event_desc_t *event_desc;
+    ucc_queue_elem_t *elem;
+
+    if (ucc_queue_is_empty(queue)) {
+        return UCC_ERR_NOT_FOUND;
+    }
+
+    ucc_spin_lock(&ee->lock);
+    elem = ucc_queue_pull(queue);
```

Comment:
do you intend to support multithread there? if yes, then isn't this a race:
threads A and B check if ucc_queue_is_empty: there is 1 entry both threads see it, and proceed. Then thread A grabs the lock, takes the element from queue and releases the lock. Now queue is empty: thread B pulls from empty queue.

Reply:
you are right. queue empty check also under spinlock

Reply:
fixed

Reply:
maybe even better with "double evaluation": first check queue is epmty w/o the lock. then go to your lines 282-286, ie. check again with the lock. This way you can amortize locking when queue is empty. Ideally we should think of some better lock-free queue or smth: this looks very similar to our progress queue, we might have a contention there. But this is for later ..